lib/rules/no-extra-parens.js

```diff
@@ -321,6 +394,11 @@ module.exports = function(context) {
                         // RegExp literal is allowed to have parens (#1589)
                         (node.object.type === "Literal" && node.object.regex)
                     )
+                ) &&
+                !(
+                    node.object.type === "FunctionExpression" &&
```

Comment:
`FunctionExpression` becomes `FunctionDeclaration` if it was not enclosed by parens, so it needs parens.
However, `ArrowFunctionExpression` is not so :)